### DISCUSSION TITLE
Nvim, ssh, kitty and TTY

### DISCUSSION THREAD

**Discussion by gzagatti** [kovidgoyal/kitty#5891] [General]
I am accessing a server using ssh with:
```
kitty +kitten ssh myserver
```

I log into the server, all kitty commands work. 

I start `nvim` and attempt:
```
:!kitty @select-window --self
```
which raises:
```
Traceback (most recent call last):                                                       
  File "runpy.py", line 197, in _run_module_as_main                                      
  File "runpy.py", line 87, in _run_code                                                 
  File "kitty_main.py", line 7, in <module>                                              
  File "kitty/entry_points.py", line 206, in main                                        
  File "kitty/entry_points.py", line 23, in remote_control                               
  File "kitty/remote_control.py", line 504, in main                                      
  File "kitty/remote_control.py", line 314, in do_io                                     
  File "kitty/utils.py", line 528, in __enter__                                          
OSError: Failed to open controlling terminal: /dev/tty (identified with ctermid()) with error: No such device or address                                                          
                                                                                         
shell returned 1
```

Similarly, attempting to run the following command:
```
:!echo 'foo' | kitty +kitten clipboard 
```
raises:
```
Traceback (most recent call last):                                                       
  File "runpy.py", line 197, in _run_module_as_main                                      
  File "runpy.py", line 87, in _run_code                                                 
  File "kitty_main.py", line 7, in <module>                                              
  File "kitty/entry_points.py", line 201, in main                                        
  File "kitty/entry_points.py", line 152, in namespaced                                  
  File "kitty/entry_points.py", line 133, in run_kitten                                  
  File "kittens/runner.py", line 112, in run_kitten                                      
  File "runpy.py", line 213, in run_module                                               
  File "runpy.py", line 87, in _run_code                                                 
  File "kittens/clipboard/main.py", line 147, in <module>                                
  File "kittens/clipboard/main.py", line 126, in main                                    
OSError: [Errno 6] No such device or address: '/dev/tty'                                 
                                                                                         
shell returned 1 
```   

Any ideas on how to solve this issue?

**Comment by kovidgoyal** [comment]
kitty @ requires either a TTY device or a socket. You need to get nvim
to run kitty @ connected to the tty device.

**Reply by gzagatti** [reply]
How do I do that? In my local machine I've never experienced such an issue.

**Reply by kovidgoyal** [reply]
Because in your local machine you are presumably using a socket. You will need to forward that over ssh. As for using the tty device in nvim, I have no idea, I dont use nvim, some googling will show you how.

**Reply by gzagatti** [reply]
Thanks. Unfortunately, I'm not quite familiar with these terms, so they are not so actionable.

I have made some experiments, running the same commands on vi or vim produce the desired effects. So the problem must be with neovim.

Now, I'd need to figure out why neovim has the expected behaviour in my local machine but does not have the same behaviour over ssh. I'll have to do some googling as you mentioned. 

Any help in this direction would be very helpful.

**Comment by gzagatti** [comment]
I have found a way around to copy from neovim using this plugin: https://github.com/ojroques/vim-oscyank

It's not quite a full solution, but at least it allows me to get text from neovim and paste in another window.

**Comment by gzagatti** [comment]
So I think I have finally solved this issue and figured what you meant by:

> kitty @ requires either a TTY device or a socket

The discussion thread #3321 led me in the right direction, in particular @groves detailed [answer](https://github.com/kovidgoyal/kitty/discussions/3321#discussioncomment-3255377). The link to the [manual entry on the rc protocol](https://sw.kovidgoyal.net/kitty/rc_protocol/) made much more sense after reading his answer. Although there is a lot of overlap between both.

I also found out that Neovim does not spawn programs invoked with the `!` connected to Neovim tty. Thread neovim/neovim#1496 documents this issue. Therefore, it's no surprise that the command `:!kitty @select-window --self` does not work over `ssh` and raises the error about the lack of a tty device. A command like `:!echo hi > /dev/tty` also does not work in Neovim both over ssh and locally.

With that being said it's then surprising to me that `:!kitty @select-window --self` works on my local machine.

Moving on to sockets. I figure that I could modify the plugins that involved kitty and were not working remotely by using a socket. First, I bound a remote socket to the kitty socket in the local machine by invoking ssh with `-R remote_socket:local_kitty_socket`. Next, I ensure that all kitty commands issued from Neovim are routed through the socket. In that way, the commands work both locally and remotely.

**Reply by AniAggarwal** [reply]
Could you perhaps share your relevant changes/config? I'm not able to get the socket forwarding to work :'(

**Reply by kovidgoyal** [reply]
Remote control forwarding is builtin to the ssh kitten these days, you just need to enable it: https://sw.kovidgoyal.net/kitty/kittens/ssh/#opt-kitten-ssh.forward_remote_control

**Reply by AniAggarwal** [reply]
I saw that but I'm unfortunately not able to get it to work. Sorry for asking to be spoonfed but I'm not able to figure out how to get this to work. I have issues on this setup:
local machine:
- mac ( :vomiting_face: ) with latest kitty installed
- have `allow_remote_control yes`, `listen_on unix:@/tmp/mykitty` (switching this to tcp localhost 0 doesn't seem to change behavior)
- have `forward_remote_control yes` set in `ssh.conf`. Also have tried removing ssh.conf and passing in the argument as `kitten ssh --kitten forward_remote_control=yes hostname`. 

remote machine:
- amazon linux 2
- uses GLIBC 2.26 (too old for anything but 0.23 kitty, I believe, which is installed. cannot be upgraded).
- installed neovim here, attempting to use https://github.com/mrjones2014/smart-splits.nvim. It successfully detects kitty backend but doesn't allow me to go from inside neovim across a kitty panel. I know that is not your project but perhaps if there is something obviously wrong with my configuration with kitty you could point that out?

**Comment by kovidgoyal** [comment]
Do

kitty -o allow_remote_control=socket-only -o listen_on=unix:/tmp/mykitty

then in kitty do,

kitten ssh localhost

then inside the ssh prompt

run

kitten @ ls

if you get output, ssh forwarding is working. Verify with
env | grep KITTY

you should see KITTY_LISTEN_ON env var set.

Now repeat with remote host instead of localhost.

Once you verify that forwarding is working, you can proceed to debug
neovim+plugins. Cant help you with that.

**Reply by AniAggarwal** [reply]
Sorry, took me a while to get an ssh server set up as I had disabled for security. Am now able to ssh to local host on my personal linux laptop and get the following output: `KITTY_LISTEN_ON=tcp:localhost:37493`.

Will verify later today on work mac and remote host (as I cannot connect to remote machine via personal laptop) and hopefully get same results. Thanks for the helpful debugging steps!

**Reply by AniAggarwal** [reply]
I have verified it works on my mac on the local host, but unfortunately doesn't seem to on the remote machine.
Both running `kitten ssh remote` and `kitten ssh --kitten forward_remote_control=yes remote` don't seem to work. The output of `env | grep kitty`: 
```bash
TERM=xterm-kitty
PATH= ... /home/aniagg/.local/share/kitty-ssh-kitten/kitty/bin: ...
OTHER_INTERN_PATH_THING=... :/usr/bin:/home/aniagg/.local/share/kitty-ssh-kitten/kitty/bin ...
```
I have verified that `kitten @ ls` works and outputs some info.

**Comment by kovidgoyal** [comment]
Then your remote machine has probably disabled ssh port forwarding in
its sshd config.

**Reply by AniAggarwal** [reply]
Interestingly the sshd_config file has these lines in it:
```
# Enabled to allow onward sessions
AllowAgentForwarding yes

# Other forwarding options for this host
X11Forwarding yes
X11DisplayOffset 10
AllowTcpForwarding yes
PermitTunnel no
```

Running this command results in these relevant lines:
```
$ kitten ssh -v --kitten forward_remote_control=yes remote_host

debug1: Requesting forwarding of remote forward LOCALHOST:0 -> /tmp/mykitty-11123:-2
debug1: mux_confirm_remote_forward: success for: listen 0, connect /tmp/mykitty-11123:-2
debug1: Allocated port 35631 for mux remote forward to (null):-2
debug1: Updating allowed port 35631 for forwarding to host /tmp/mykitty-11123 port -2
Allocated port 35631 for remote forward to :-2
```

EDIT: I just noticed that if I do not pass the `forward_remote_control=yes` flag when using kitten ssh, my smart splits plugin does not allow me to move within neovim windows and doesn't detect a backend in the logs (using IS_NVIM env var I think), but when passing that flag it does detect kitty backend but instead allows me to move within neovim but not across kitty windows.

**Reply by kovidgoyal** [reply]
I am confused. You said you dont have KITTY_LISTEN_ON but kitten @ ls still works? How is that possible? Did you set allow_remote_control to socket-only?

**Reply by AniAggarwal** [reply]
Not sure how that works. In my kitty config I have `allow_remote_control yes` but I followed these setps to get the below output:

1. launched a new kitty os window via `kitty -o allow_remote_control=socket-only -o listen_on=unix:/tmp/mykitty`
2. `kitten ssh --kitten forward_remote_control=yes remote_host`
3. `kitten @ ls` works
4. `env | grep kitty` does not show `KITTY_LISTEN_ON`.
5. Run `kitten ssh -v --kitten forward_remote_control=yes remote_host` for below output (only some lines included:
```
debug1: Requesting forwarding of remote forward LOCALHOST:0 -> /tmp/mykitty-17311:-2
Allocated port 43725 for remote forward to :-2
```

**Reply by AniAggarwal** [reply]
One thing I thought of that could be the root of the issue is that on the remote machine the kitten located in `~/.local/share/kitty-ssh-kitten/kitty/bin/` runs fine and is version 0.35. However, running the kitty in that same directory results in a glib issue:
```
./kitty --version 
/local/home/aniagg/.local/share/kitty-ssh-kitten/kitty/install/bin/kitty: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by /local/home/aniagg/.local/share/kitty-ssh-kitten/kitty/install/bin/../lib/libpython3.12.so.1.0)
```
Could this be causing issues?
The latest version of kitty I can have installed on the remote machine is `0.23.1`, which runs fine but is obviously too old to support these cool new features. I can't upgrade glib from 2.26 unfortunately.

**Comment by kovidgoyal** [comment]
No, kitty is not used anymore on remote machines. kitten does it all.
In this case even kitten is used only on the local computer not the
remote one. Port forwarding is done by ssh and the env var is set by
the bootstrap shell script.

**Reply by AniAggarwal** [reply]
Got it. It seems like the port is be allocated correctly, so I'm not sure what's going wrong or why the KITTY_LISTEN_ON variable isn't set. Perhaps this is the issue then?

I also just confirmed that running the below steps on my local machine ssh'd back to local machine has the nvim + plugins all work as expected, so we can hopefully rule those out for now, at least. 
1. launched a new kitty os window via `kitty -o allow_remote_control=socket-only -o listen_on=unix:/tmp/mykitty`
2. `kitten ssh --kitten forward_remote_control=yes localhost`
3. `kitten @ ls` works
4. `env | grep kitty` shows `KITTY_LISTEN_ON`.
5. Running neovim with smart splits plugins all works correctly. Kitty backend discovered, able to move across splits in neovim and in kitty.

**Comment by kovidgoyal** [comment]
On Mon, Jun 24, 2024 at 08:12:28AM -0700, Ani Aggarwal wrote:
> Got it. It seems like the port is be allocated correctly, so I'm not sure what's going wrong or why the KITTY_LISTEN_ON variable isn't set. Perhaps this is the issue then?

Most likely something in the remote server environment is resetting env
vars. Try using a empty shell rc file or a different shell.

**Reply by AniAggarwal** [reply]
OH I was being very stupid. I ran `env | grep kitty` all this time instead of `env | grep KITTY`. Looks like KITTY_LISTEN_ON is set even without using an empty shell rc. I get `KITTY_LISTEN_ON=tcp:localhost:39767`.
Is it safe to assume this isn't a kitty issue then and must be with neovim/plugins?
I assume the user vars are all being set properly if KITTEN_LISTEN_ON is set?

**Comment by kovidgoyal** [comment]
If the env var is set and kitten @ is working then port forwarding is
working. User vars dont have anything to do with port forwarding, they
are set via escape codes on the tty itself.
